File: 67.add-binary.cpp
/*
 * @lc app=leetcode id=67 lang=cpp
 *
 * [67] Add Binary
 */

#include <algorithm>
#include <string>

using namespace std;

// @lc code=start
class Solution {
public:
    string addBinary(string a, string b) {
        string r;
        if (a.length() > b.length()) {
            swap(a, b);
        }
        int carry = 0;
        auto i = 0;
        for (; i < a.length(); ++i) {
            auto ac = (int)(a[a.size() - i - 1] - '0');
            auto bc = (int)(b[b.size() - i - 1] - '0');
            auto s = ac + bc + carry;
            carry = s >> 1;
            s = s & 1;
            r.push_back((char)(s + '0'));
        }
        for (; i < b.length(); ++i) {
            auto bc = (int)(b[b.size() - i - 1] - '0');
            auto s = bc + carry;
            carry = s >> 1;
            s = s & 1;
            r.push_back((char)(s + '0'));
        }
        if (carry == 1) {
            r.push_back('1');
        }
        reverse(r.begin(), r.end());
        return r;
    }
};
// @lc code=end

File: 67.add-binary.py
#
# @lc app=leetcode id=67 lang=python3
#
# [67] Add Binary
#

# @lc code=start
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        carry = False
        l, r = (a, b) if len(a) > len(b) else (b, a)
        dl = len(l) - len(r)
        r = '0' * dl + r
        print(l)
        print(r)
        res = ''
        for i in reversed(range(len(l))):
            lc, rc = (l[i], r[i]) if l[i] == '0' else (r[i], l[i])
            if lc == '0':
                if rc == '0':
                    res += ('1' if carry else '0')
                    carry = False
                else:
                    res += ('0' if carry else '1')
            else:
                res += ('1' if carry else '0')
                carry = True
        if carry:
            res += '1'
        return res[::-1]
# @lc code=end

from test_utils import run_test

method = Solution().addBinary

run_test(method, ['11', '1'], '100')
run_test(method, ['1010', '1011'], '10101')
